packages/engine/paima-sm/src/delegate-wallet.ts

```diff
@@ -0,0 +1,308 @@
+    },
+  };
+
+  private static parser = new PaimaParser(
+    DelegateWallet.delegationGrammar,
+    DelegateWallet.parserCommands
+  );
+
+  constructor(private DBConn: PoolClient) {}
+
+  /* Generate Plaintext Message */
+  private generateMessage(internalMessage: string = ''): string {
+    return `${DelegateWallet.DELEGATE_WALLET_PREFIX}${
+      DelegateWallet.SEP
+    }${internalMessage.toLocaleLowerCase()}${DelegateWallet.SEP}${ENV.CONTRACT_ADDRESS}`;
+  }
+
+  /* Verify Signature with all possible wallets. */
+  private async verifySignature(
+    walletAddress: string,
+    message: string,
+    signature: string
+  ): Promise<boolean> {
+    if (!walletAddress || !signature) throw new Error('No Signature');
+    const wallets = [
+      CryptoManager.Evm(new Web3()),
+      CryptoManager.Algorand(),
+      CryptoManager.Cardano(),
+      CryptoManager.Polkadot(),
+    ];
+    for (const wallet of wallets) {
+      try {
+        if (await wallet.verifySignature(walletAddress, message, signature)) {
+          return true;
+        }
+      } catch {
+        // do nothing.
+      }
+    }
+    throw new Error(`Invalid Signature for ${walletAddress} : ${message}`);
+  }
+
+  public async createAddress(address: string): Promise<IGetAddressFromAddressResult> {
+    await newAddress.run({ address: address }, this.DBConn);
+    const [createdAddress] = await getAddressFromAddress.run({ address }, this.DBConn);
+    return createdAddress;
+  }
+
+  public async getAddress(address: string): Promise<AddressWrapper> {
+    const [exitingAddress] = await getAddressFromAddress.run({ address: address }, this.DBConn);
+    if (!exitingAddress) return { address, exists: false };
+    return { address: exitingAddress, exists: true };
+  }
+
+  // Swap:
+  // addressB gains addressA ID.
+  // addressA is assigned a new ID
+  private async swap(
+    existingAddress: AddressExists,
+    nonExistingAddress: AddressDoesNotExist
+  ): Promise<[AddressExists, AddressExists]> {
+    await updateAddress.run(
+      {
+        new_address: nonExistingAddress.address,
+        old_address: existingAddress.address.address,
+      },
+      this.DBConn
+    );
+    await newAddress.run({ address: existingAddress.address.address }, this.DBConn);
+
+    return await Promise.all([
+      this.getAddress(existingAddress.address.address) as Promise<AddressExists>,
+      this.getAddress(nonExistingAddress.address) as Promise<AddressExists>,
+    ]);
+  }
+
+  private async validateDelegate(
+    fromAddress: AddressWrapper,
+    toAddress: AddressWrapper
+  ): Promise<void> {
+    if (!fromAddress.exists || !toAddress.exists) {
+      return; // both are new.
+    }
+    if (fromAddress.address.id === toAddress.address.id) {
+      throw new Error('Cannot delegate to itself');
+    }
+    if (fromAddress.exists && toAddress.exists) {
+      throw new Error('Both A and B have progress. Cannot merge.');
+    }
+
+    await Promise.all([
```

Comment:
Probably the code will be a lot cleaner if instead of checking the error message inside the promise, you return a tuple of 3 value and then throw the errors afterwards